eventtracking/backends/event_bus.py

```diff
@@ -0,0 +1,59 @@
+"""Event tracker backend that emits events to the event-bus."""
+
+import json
+import logging
+from datetime import datetime
+
+from django.conf import settings
+from openedx_events.analytics.data import TrackingLogData
+from openedx_events.analytics.signals import TRACKING_EVENT_EMITTED
+from openedx_events.event_bus import get_producer
+
+from eventtracking.backends.routing import RoutingBackend
+from eventtracking.config import SEND_TRACKING_EVENT_EMITTED_SIGNAL
+
+logger = logging.getLogger(__name__)
+
+
+class EventBusRoutingBackend(RoutingBackend):
+    """
+    Event tracker backend for the event bus.
+    """
+
+    def send(self, event):
+        """
+        Send the tracking log event to the event bus by emitting the
+        TRACKING_EVENT_EMITTED signal using custom metadata.
+        """
+        if not SEND_TRACKING_EVENT_EMITTED_SIGNAL.is_enabled():
+            return
+
+        name = event.get("name")
+
+        if name not in getattr(settings, "EVENT_BUS_TRACKING_LOGS", []):
+            return
+
+        data = json.dumps(event.get("data"))
+        context = json.dumps(event.get("context"))
+
+        timestamp = event.get("timestamp")
+
+        if isinstance(timestamp, str):
+            timestamp = datetime.strptime(timestamp, "%Y-%m-%dT%H:%M:%S.%f%z")
+
+        tracking_log = TrackingLogData(
+            name=event.get("name"),
+            timestamp=timestamp,
+            data=data,
+            context=context,
+        )
+        metadata = TRACKING_EVENT_EMITTED.generate_signal_metadata()
+        get_producer().send(
+            signal=TRACKING_EVENT_EMITTED,
+            topic="analytics",
+            event_key_field="tracking_log.name",
+            event_data={"tracking_log": tracking_log},
+            event_metadata=metadata,
+        )
```

Comment:
Will that work if the signal is consumed on the same runtime as the `lms`?

I had to generate the metadata and call `get_producer` to avoid infinite recursion.

Reply:
The consumer would be a separate runtime. We'd just have to ensure that it wasn't configured to re-produce the event (which I think should be the default case).

In the general case we'd like to implement https://github.com/openedx/openedx-events/issues/79 to prevent that sort of issue, though.

Reply:
In any case, get_producer should work fine for now -- it's just more code to maintain, and a possible later risk if there needs to be a breaking change to the producer API of openedx-events.

Reply:
Tested locally, events are processed again and again in `event-routing-backends` indefinetely. Note the ID in the transformer output, it's idempotent and it's repeated for the play_video tracking log:

```
2024-02-08 17:31:32,204 INFO 1091 [edx_event_bus_redis.internal.producer] [user None] [ip None] producer.py:86 - Message delivered to Redis event bus: topic=dev-analytics, message_id=db434498-c6a7-11ee-8735-0242ac12000f, signal=<OpenEdxPublicSignal: org.openedx.analytics.tracking.event.emitted.v1>, redis_msg_id=b'1707413492203-0'
2024-02-08 17:31:32,205 INFO 1091 [edx_event_bus_redis.internal.consumer] [user None] [ip None] consumer.py:296 - Message from Redis processed successfully
2024-02-08 17:31:32,208 INFO 1091 [edx_event_bus_redis.internal.consumer] [user None] [ip None] consumer.py:350 - Message received from Redis: topic=dev-analytics, message_id=db54ddd4-c6a7-11ee-a06a-0242ac12000f, redis_msg_id=b'1707413492127-0', event_timestamp_ms=1707413473.66753
2024-02-08 17:31:32,208 INFO 1091 [event_routing_backends.handlers] [user None] [ip None] handlers.py:54 - Successfully processed event "pause_video"
2024-02-08 17:31:32,251 WARNING 1091 [event_routing_backends.processors.mixins.base_transformer] [user None] [ip None] base_transformer.py:179 - Could not get value for data.current_time in event "pause_video"
2024-02-08 17:31:32,252 INFO 1091 [xapi_tracking] [user None] [ip None] transformer_processor.py:61 - {"id": "59414c48-929a-538c-8b74-6bfe3b9f098a", "result": {"extensions": {"https://w3id.org/xapi/video/extensions/time": 35.178}}, "version": "1.0.3", "actor": {"objectType": "Agent", "account": {"name": "cde636ff-7f3a-4170-bbe4-b5ee99df5946", "homePage": "http://local.overhang.io:8000"}}, "verb": {"id": "https://w3id.org/xapi/video/verbs/paused", "display": {"en": "paused"}}, "object": {"id": "http://local.overhang.io:8000/xblock/block-v1:edunext+demo+1+type@video+block@fd29b1bd111643a78052edc9260329ab", "objectType": "Activity", "definition": {"type": "https://w3id.org/xapi/video/activity-type/video"}}, "timestamp": "2024-02-08T17:31:13.662932+00:00", "context": {"contextActivities": {"parent": [{"id": "http://local.overhang.io:8000/course/course-v1:edunext+demo+1", "objectType": "Activity", "definition": {"name": {"en-US": "Demo"}, "type": "http://adlnet.gov/expapi/activities/course"}}]}, "extensions": {"https://w3id.org/xapi/openedx/extension/transformer-version": "event-routing-backends@8.0.0", "https://w3id.org/xapi/openedx/extensions/session-id": "b2a9e6be2497d9ca776ebbbc36a0e808", "https://w3id.org/xapi/video/extensions/length": 141.0}}}
2024-02-08 17:31:32,284 INFO 1091 [edx_event_bus_redis.internal.producer] [user None] [ip None] producer.py:86 - Message delivered to Redis event bus: topic=dev-analytics, message_id=db54ddd4-c6a7-11ee-a06a-0242ac12000f, signal=<OpenEdxPublicSignal: org.openedx.analytics.tracking.event.emitted.v1>, redis_msg_id=b'1707413492284-0'
2024-02-08 17:31:32,285 INFO 1091 [edx_event_bus_redis.internal.consumer] [user None] [ip None] consumer.py:296 - Message from Redis processed successfully
2024-02-08 17:31:32,288 INFO 1091 [edx_event_bus_redis.internal.consumer] [user None] [ip None] consumer.py:350 - Message received from Redis: topic=dev-analytics, message_id=db434498-c6a7-11ee-8735-0242ac12000f, redis_msg_id=b'1707413492203-0', event_timestamp_ms=1707413473.552267
2024-02-08 17:31:32,289 INFO 1091 [event_routing_backends.handlers] [user None] [ip None] handlers.py:54 - Successfully processed event "play_video"
2024-02-08 17:31:32,330 WARNING 1091 [event_routing_backends.processors.mixins.base_transformer] [user None] [ip None] base_transformer.py:179 - Could not get value for data.current_time in event "play_video"
2024-02-08 17:31:32,332 INFO 1091 [xapi_tracking] [user None] [ip None] transformer_processor.py:61 - {"id": "afc758ed-53e8-5890-8cd6-68e7c9ed595f", "result": {"extensions": {"https://w3id.org/xapi/video/extensions/time": 35.071}}, "version": "1.0.3", "actor": {"objectType": "Agent", "account": {"name": "cde636ff-7f3a-4170-bbe4-b5ee99df5946", "homePage": "http://local.overhang.io:8000"}}, "verb": {"id": "https://w3id.org/xapi/video/verbs/played", "display": {"en": "played"}}, "object": {"id": "http://local.overhang.io:8000/xblock/block-v1:edunext+demo+1+type@video+block@fd29b1bd111643a78052edc9260329ab", "objectType": "Activity", "definition": {"type": "https://w3id.org/xapi/video/activity-type/video"}}, "timestamp": "2024-02-08T17:31:13.544318+00:00", "context": {"contextActivities": {"parent": [{"id": "http://local.overhang.io:8000/course/course-v1:edunext+demo+1", "objectType": "Activity", "definition": {"name": {"en-US": "Demo"}, "type": "http://adlnet.gov/expapi/activities/course"}}]}, "extensions": {"https://w3id.org/xapi/openedx/extension/transformer-version": "event-routing-backends@8.0.0", "https://w3id.org/xapi/openedx/extensions/session-id": "b2a9e6be2497d9ca776ebbbc36a0e808", "https://w3id.org/xapi/video/extensions/length": 141.0}}}
2024-02-08 17:31:32,363 INFO 1091 [edx_event_bus_redis.internal.producer] [user None] [ip None] producer.py:86 - Message delivered to Redis event bus: topic=dev-analytics, message_id=db434498-c6a7-11ee-8735-0242ac12000f, signal=<OpenEdxPublicSignal: org.openedx.analytics.tracking.event.emitted.v1>, redis_msg_id=b'1707413492362-0'
2024-02-08 17:31:32,363 INFO 1091 [edx_event_bus_redis.internal.consumer] [user None] [ip None] consumer.py:296 - Message from Redis processed successfully
2024-02-08 17:31:32,366 INFO 1091 [edx_event_bus_redis.internal.consumer] [user None] [ip None] consumer.py:350 - Message received from Redis: topic=dev-analytics, message_id=db54ddd4-c6a7-11ee-a06a-0242ac12000f, redis_msg_id=b'1707413492284-0', event_timestamp_ms=1707413473.66753
2024-02-08 17:31:32,367 INFO 1091 [event_routing_backends.handlers] [user None] [ip None] handlers.py:54 - Successfully processed event "pause_video"
2024-02-08 17:31:32,405 WARNING 1091 [event_routing_backends.processors.mixins.base_transformer] [user None] [ip None] base_transformer.py:179 - Could not get value for data.current_time in event "pause_video"
2024-02-08 17:31:32,406 INFO 1091 [xapi_tracking] [user None] [ip None] transformer_processor.py:61 - {"id": "59414c48-929a-538c-8b74-6bfe3b9f098a", "result": {"extensions": {"https://w3id.org/xapi/video/extensions/time": 35.178}}, "version": "1.0.3", "actor": {"objectType": "Agent", "account": {"name": "cde636ff-7f3a-4170-bbe4-b5ee99df5946", "homePage": "http://local.overhang.io:8000"}}, "verb": {"id": "https://w3id.org/xapi/video/verbs/paused", "display": {"en": "paused"}}, "object": {"id": "http://local.overhang.io:8000/xblock/block-v1:edunext+demo+1+type@video+block@fd29b1bd111643a78052edc9260329ab", "objectType": "Activity", "definition": {"type": "https://w3id.org/xapi/video/activity-type/video"}}, "timestamp": "2024-02-08T17:31:13.662932+00:00", "context": {"contextActivities": {"parent": [{"id": "http://local.overhang.io:8000/course/course-v1:edunext+demo+1", "objectType": "Activity", "definition": {"name": {"en-US": "Demo"}, "type": "http://adlnet.gov/expapi/activities/course"}}]}, "extensions": {"https://w3id.org/xapi/openedx/extension/transformer-version": "event-routing-backends@8.0.0", "https://w3id.org/xapi/openedx/extensions/session-id": "b2a9e6be2497d9ca776ebbbc36a0e808", "https://w3id.org/xapi/video/extensions/length": 141.0}}}
```

Note that I'm using the following command to trigger the consumption of the events:

```
tutor dev exec lms ./manage.py lms consume_events -t analytics -g event_routing_backends --extra '{"consumer_name": "aspects"}'
```

And this is my lms settings:

```python
EVENT_BUS_PRODUCER_CONFIG = {
     "org.openedx.analytics.tracking.event.emitted.v1": {
        "analytics": {"event_key_field": "tracking_log.name", "enabled": True}
        }
      }
```

Reply:
Yes, the consumer would have to not have that event enabled in its `EVENT_BUS_PRODUCER_CONFIG`, otherwise you'll get that loop.

I don't know if we'll work on https://github.com/openedx/openedx-events/issues/79 any time soon, but if someone is interested in tackling it I don't think there would be any blockers to that.

Reply:
@timmc-edx this is what I come up with: https://github.com/openedx/openedx-events/pull/312

Can you take a look?